File: agriml_system/cpp_weather/kafka_client.cpp
// Kafka client integration for C++ weather module using librdkafka

#include <iostream>
#include <string>
#include <librdkafka/rdkafkacpp.h>

class KafkaClient {
public:
    KafkaClient(const std::string& brokers, const std::string& group_id) {
        std::string errstr;

        // Producer config
        RdKafka::Conf* conf = RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL);
        conf->set("bootstrap.servers", brokers, errstr);

        producer_ = RdKafka::Producer::create(conf, errstr);
        if (!producer_) {
            std::cerr << "Failed to create producer: " << errstr << std::endl;
            exit(1);
        }

        // Consumer config
        RdKafka::Conf* tconf = RdKafka::Conf::create(RdKafka::Conf::CONF_TOPIC);
        consumer_conf_ = RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL);
        consumer_conf_->set("group.id", group_id, errstr);
        consumer_conf_->set("bootstrap.servers", brokers, errstr);

        consumer_ = RdKafka::KafkaConsumer::create(consumer_conf_, errstr);
        if (!consumer_) {
            std::cerr << "Failed to create consumer: " << errstr << std::endl;
            exit(1);
        }
    }

    ~KafkaClient() {
        delete producer_;
        delete consumer_;
        delete consumer_conf_;
    }

    void send_message(const std::string& topic, const std::string& message) {
        RdKafka::ErrorCode resp = producer_->produce(
            topic, RdKafka::Topic::PARTITION_UA,
            RdKafka::Producer::RK_MSG_COPY,
            const_cast<char*>(message.c_str()), message.size(),
            nullptr, nullptr);

        if (resp != RdKafka::ERR_NO_ERROR) {
            std::cerr << "Failed to send message: " << RdKafka::err2str(resp) << std::endl;
        } else {
            producer_->poll(0);
        }
    }

    void consume_messages(const std::string& topic) {
        std::vector<std::string> topics = {topic};
        RdKafka::ErrorCode err = consumer_->subscribe(topics);
        if (err) {
            std::cerr << "Failed to subscribe to topic: " << RdKafka::err2str(err) << std::endl;
            return;
        }

        while (true) {
            RdKafka::Message* msg = consumer_->consume(1000);
            if (!msg) continue;

            switch (msg->err()) {
                case RdKafka::ERR__TIMED_OUT:
                    break;
                case RdKafka::ERR_NO_ERROR:
                    std::cout << "Received message: " << static_cast<const char*>(msg->payload()) << std::endl;
                    break;
                default:
                    std::cerr << "Consume failed: " << msg->errstr() << std::endl;
                    break;
            }
            delete msg;
        }
    }

private:
    RdKafka::Producer* producer_;
    RdKafka::KafkaConsumer* consumer_;
    RdKafka::Conf* consumer_conf_;
};

File: agriml_system/rust_edge/src/kafka_client.rs
// Kafka client integration for Rust edge module using rdkafka

use rdkafka::config::ClientConfig;
use rdkafka::producer::{FutureProducer, FutureRecord};
use rdkafka::consumer::{StreamConsumer, Consumer};
use rdkafka::message::Message;
use std::time::Duration;
use futures::StreamExt;

pub struct KafkaClient {
    producer: FutureProducer,
    consumer: StreamConsumer,
}

impl KafkaClient {
    pub fn new(brokers: &str, group_id: &str) -> KafkaClient {
        let producer: FutureProducer = ClientConfig::new()
            .set("bootstrap.servers", brokers)
            .create()
            .expect("Producer creation error");

        let consumer: StreamConsumer = ClientConfig::new()
            .set("group.id", group_id)
            .set("bootstrap.servers", brokers)
            .set("enable.partition.eof", "false")
            .set("session.timeout.ms", "6000")
            .set("enable.auto.commit", "true")
            .create()
            .expect("Consumer creation failed");

        KafkaClient { producer, consumer }
    }

    pub async fn send_message(&self, topic: &str, payload: &str) {
        let record = FutureRecord::to(topic)
            .payload(payload)
            .key("");
        let _ = self.producer.send(record, Duration::from_secs(0)).await;
    }

    pub async fn consume_messages(&self, topic: &str) {
        self.consumer.subscribe(&[topic]).expect("Subscription failed");
        let mut message_stream = self.consumer.stream();

        while let Some(message) = message_stream.next().await {
            match message {
                Ok(m) => {
                    if let Some(payload) = m.payload_view::<str>().unwrap_or(None) {
                        println!("Received message: {}", payload);
                    }
                }
                Err(e) => eprintln!("Kafka error: {}", e),
            }
        }
    }
}
